quil-py/src/program/calibration.rs

```diff
@@ -0,0 +1,134 @@
+use pyo3::prelude::*;
+use quil_rs::{
+    expression::Expression,
+    instruction::{
+        Calibration, GateModifier, Instruction, MeasureCalibrationDefinition, Measurement, Qubit,
+    },
+    program::CalibrationSet,
+};
+use rigetti_pyo3::{
+    impl_as_mut_for_wrapper, impl_repr, py_wrap_type, PyTryFrom, PyWrapper, PyWrapperMut, ToPython,
+    ToPythonError,
+};
+
+use crate::instruction::{
+    PyCalibration, PyExpression, PyGateModifier, PyInstruction, PyMeasureCalibrationDefinition,
+    PyMeasurement, PyQubit,
+};
+
+use super::ProgramError;
+
+py_wrap_type! {
+    PyCalibrationSet(CalibrationSet) as "CalibrationSet"
+}
+impl_as_mut_for_wrapper!(PyCalibrationSet);
+impl_repr!(PyCalibrationSet);
+
+#[pymethods]
+impl PyCalibrationSet {
+    #[new]
+    pub fn new(
+        py: Python<'_>,
+        calibrations: Vec<PyCalibration>,
+        measure_calibrations: Vec<PyMeasureCalibrationDefinition>,
+    ) -> PyResult<Self> {
+        Ok(Self(CalibrationSet {
+            calibrations: Vec::<_>::py_try_from(py, &calibrations)?,
+            measure_calibrations: Vec::<_>::py_try_from(py, &measure_calibrations)?,
+        }))
+    }
+
+    #[getter]
+    pub fn calibrations(&self, py: Python<'_>) -> PyResult<Vec<PyCalibration>> {
+        self.as_inner().calibrations().to_python(py)
+    }
+
+    #[getter]
+    pub fn measure_calibrations(
+        &self,
+        py: Python<'_>,
+    ) -> PyResult<Vec<PyMeasureCalibrationDefinition>> {
+        self.as_inner().measure_calibrations().to_python(py)
+    }
+
+    pub fn expand(
+        &self,
+        py: Python<'_>,
+        instruction: PyInstruction,
+        previous_calibrations: Vec<PyInstruction>,
+    ) -> PyResult<Option<Vec<PyInstruction>>> {
+        self.as_inner()
+            .expand(
+                &Instruction::py_try_from(py, &instruction)?,
+                &Vec::<Instruction>::py_try_from(py, &previous_calibrations)?,
+            )
+            .map_err(ProgramError::from)
+            .map_err(ProgramError::to_py_err)?
+            .to_python(py)
+    }
+
+    pub fn get_match_for_measurement(
+        &self,
+        py: Python<'_>,
+        measurement: PyMeasurement,
+    ) -> PyResult<Option<PyMeasureCalibrationDefinition>> {
+        Ok(self
+            .as_inner()
+            .get_match_for_measurement(&Measurement::py_try_from(py, &measurement)?)
+            .map(PyMeasureCalibrationDefinition::from))
+    }
+
+    pub fn get_match_for_gate(
+        &self,
+        py: Python<'_>,
+        gate_modifiers: Vec<PyGateModifier>,
+        gate_name: &str,
+        gate_parameters: Vec<PyExpression>,
+        gate_qubits: Vec<PyQubit>,
+    ) -> PyResult<Option<PyCalibration>> {
```

Comment:
(Corresponding to [this comment](https://github.com/rigetti/pyquil/pull/1536#discussion_r1132781614))

Two things here:

* `get_match_for_measurement` takes a measurement as a single param, but `get_match_for_gate` requires the caller to decompose the gate. Why's that?
* ~~Could this just be `get_match_for_instruction` and do the match on this side? Keeps more of the logic together in this repo rather than farming some of it out to pyQuil~~ I see the benefit of having a narrower type in the return signature.

Reply:
Yeah, this is a question I was going to pose once this was out of draft, but worth considering now:

To explain why they are different, `get_match_for_gate` was a pre-existing method, so I just mirrored it in the bindings. `get_match_for_measurement` was tied up in `expand`, so I had to lift it out and chose to just use a `Measurement` directly, rather than decompose it.

My preference is to take the latter approach, so I would like to update `get_match_for_gate` to take a `Gate` instead of the individual components. You happened to write `get_match_for_gate`, so I'll reverse the question back on to you: Is there a benefit to having the caller have to decompose the `Gate` rather than just pass a `Gate` object in? If not, then I'll make the change.

Reply:
I refactored `get_match_for_gate` to use a `&Gate` and labeled the commit as a breaking change.